### DISCUSSION TITLE
Release process proposal

### DISCUSSION THREAD

**Discussion by ematipico** [rome/tools#2471] [Suggestions]
Few days ago [on discord](https://discord.com/channels/678763474494423051/678763474930761739/963823345504227369) I proposed to make a release of the new things we had done on the formatter (some new option and some bug fix), and that raised few question about how often  we should do releases, what to release and how can we do releases without shipping things that are still unstable/experimental.

Two different topics, two different proposals.

**Comment by ematipico** [comment]
# Release cadence

How often should we do a release? How do we handle minor releases and patch releases?

Some projects release at a certain time:
- Rust does its release every six weeks, if a feature doesn't make the cut, it's released in the next release, patch releases have high priority.
- eslint seems to do a weekly release on Friday and it's always a minor (correct me if am wrong);

Other projects have don't have a clear cadence (probably because maintained by few people or because there isn't need).

As we have a big team working on the project, I think we can make a commitment and try to stick to it, so the community
knows when to expect a new release (with potential features).

## Proposal

- patch releases of bugs that are blocking have a priority over any feature/other bugfix; how to label
a bug "blocking", I suppose they can be the ones that don't have a workaround or really blocks a developer from working;
- a four week release that will contain **not blocking fixes** and (if there are any) features;
- if a release contains new feature that worth an explanation, we should have a blog post;

**Reply by sebmck** [reply]
I like the idea of weekly releases, although rather than a Friday (which makes sense for a volunteer effort), we should target Monday or Tuesday. We may want to even plan around a major version every 6 months or so, as it may help us making long-term plans, especially around compatibility and introducing large features.

**Reply by yassere** [reply]
How much direct attention and care do we plan to put into each release? If we have a weekly release that’s fully automated, would that just be a glorified nightly?

**Reply by MichaReiser** [reply]
I like the idea of having a fixed cadence for releases. I would recommend that we align the project milestones and release cadence (by changing either of them).

**Reply by ematipico** [reply]
> I like the idea of weekly releases, although rather than a Friday (which makes sense for a volunteer effort), we should target Monday or Tuesday.

Nightly releases were good and I think we can reintroduce them once we are comfortable. I am not sure we are in a position of offering weekly releases. Each release requires some preparation from the team: update documentation on the website, changelog, possible blog post, etc. Seems lot of work. Did you have something in particular in mind?

> We may want to even plan around a major version every 6 months or so, as it may help us making long-term plans, especially around compatibility and introducing large features.

Good point! Probably major releases could be in sync with the release of new tools and (of course) breaking changes.

**Reply by NicholasLYang** [reply]
I like the idea of weekly releases for minor changes, then monthly announcement/documentation changes. That way we can have a standard "here's the cool stuff we added" post every month.

**Reply by ematipico** [reply]
Based on the feedback received, it seems that some people expressed some interested in weekly releases, although we are missing proposal on how to achieve that. Automatic? Manual? What about the changelog? 


# Final decision

- patch releases of bugs that are blocking have a priority over any feature/other bugfix; how to label
a bug "blocking", I suppose they can be the ones that don't have a workaround or really blocks a developer from working;
- a four week release that will contain non-blocking fixes and (if there are any) features; **Tuesday** will be the day of the release. Having it on a Monday doesn't seem ideal as people come from the weekend. And it will pressure the team into prepare the release on a Friday;
- if a release contains new features that's worth an explanation, we should have a blog post; this is arbitrary and it's up to the team to decide if a new minor release needs to have a blog post;

**The decisions was taken are based on the fact that we don't have an automated process to make releases**.

Once we have more time to make an automatic release (auto generation of changelog, better workflows and versioning, publish documentation only after the release, etc.), I think it's worth revisit the decision with another proposal. For example, weekly releases.

**Comment by ematipico** [comment]
# Experimental features

Having a recurring release opens us to the risk of merging new features that are not ready to be published, for different reasons, e.g. multiple PRs are needed to create the feature, performance regressions, etc.

These are features that **will eventually be merged**, regardless of the effort. **They are not experiments** that we plan
to "test" and ditch in case the "experiment" didn't go well.

So, the wording can change, we can call it "experimental", "feature flags", "WIP features", etc., but these features will eventually be the default of Rome.

## Proposal

- Having a cental crate where all tools will depend from; this is something that we have to do regardless, so starting this work sooner rather than later will benefit everyone. We can have a `rome_config` crate, or use `rome_core` that we have already. This is where tools will check if they should hide the experimental* features.
- We can leverage Rust language (macros, directives, etc.) to hide these features under the `experimental`*  feature flag.
- Having a `--experimental`* flag at CLI level that unlocks these work in progress features; all commands should support this flag.
- Having a "experimental" option in all the editors we support.
- We can different workflows to test these experimental* features.



*the name is arbitrary and not set on stone

**Reply by sebmck** [reply]
I really like how Rust handle feature flags, another precedent is [Flow](https://flow.org/en/docs/config/options/) who have experimental config options too.

**Reply by sebmck** [reply]
One deciding factor is if we choose to violate semver for these experimental features. IMO we definitely should as it would be a significant hurdle to development otherwise. Possibly considering a semver minor bump the breaking change indicator for experimental flags so people can at least pin to a minor range to get bug fixes.

**Reply by yassere** [reply]
I like the idea of unstable features that are individually opt-in. I’m not sure we need a single `experimental` flag that would enable every feature.

And I don’t believe we need to make any semver guarantees within the context of unstable features. I think that using an unstable feature can be considered analogous to using a package in the `0.y.z` range where you effectively get no stability guarantees.

**Reply by MichaReiser** [reply]
Is your proposal proposing a single solution or different alternatives? My understanding of Rust's feature flags is that the compiler produces different outputs depending on if a flag is enabled or not. That's why it won't be possible to then support a `rome --experimental` option because the experimental features are either not part of the compiled binary or always enabled. 

To clarify, are you proposing to use a single `experimental` flag or one per feature?

> One deciding factor is if we choose to violate semver for these experimental features.

I think it's in the nature of experimental features that they may change at any time and we should, therefore, not adhere to semver. How we ensure that patch releases never contain new feature is another organisational question. An option could be that we backport bugfixes to release branches and then release from the release branch.

**Reply by ematipico** [reply]
My proposal is to have **only one experimental flag across the board**. **The main proposal is more for us maintainers/contributors**, not for the user. With this proposal we would want:

- make sure that the people that work on unstable features are comfortable enough to do as much as they like without breaking existing code or behaviour;
- avoid overhead and burden on the team by maintaining multiple flags for each features they are preparing/developing;
- having a single way to introduce a new experimental feature and a single way to make them the default behaviour; once we figure out how (we may want to borrow Rust's way)

If a user decides to opt-in the experimental flag, they will do it at their own risk and they should not a expect an hotfix if something blocks them from doing their work (meaning the they may have found a bug, which is great, but we will fix it at our terms).

> My understanding of Rust's feature flags is that the compiler produces different outputs depending on if a flag is enabled or not. That's why it won't be possible to then support a rome --experimental option because the experimental features are either not part of the compiled binary or always enabled

@MichaReiser I had a chat with Léo about it, and my understanding is the following: the unstable feature are always part of the final binary, regardless, and people can use them by setting an env variable called `RUSTC_BOOTSTRAP`. By doing so, they opt-in the experimental features.

**Reply by yassere** [reply]
> make sure that the people that work on unstable features are comfortable enough to do as much as they like without breaking existing code or behaviour

I think that individual feature flags would support this goal better than a single `experimental` flag would. If you’re adding a new feature, you can confidently make changes behind a `some_new_feature` flag without worrying as much about breaking other experimental features.

We wouldn't make any explicit stability guarantees about an `experimental` flag, but if early adopters start to use it, we wouldn't want to constantly break things.

**Reply by MichaReiser** [reply]
> I think that individual feature flags would support this goal better than a single experimental flag would. If you’re adding a new feature, you can confidently make changes behind a some_new_feature flag without worrying as much about breaking other experimental features.

The challenge I see here is that multiple flags are inherently more complicated to implement and raise serious issues around testing. How do we ensure that all experimental features all play nicely together? 

But multiple flags certainly are nice if we can make it work. For example, it makes it clearer why a certain path is gated as experimental. 

What about if features have individual flags but Rome only supports switching them all ON or OFF?

**Reply by sebmck** [reply]
Having a single flag I think incentivizes people to just enable it always as it's "easy". It also makes it easy to inadvertently use other features that are behind the experimental flag without realizing it. This could lead to pretty high churn for people who don't know the contract they're implicitly signing up for when using it. It could actually increase the maintenance burden for us, and give people the wrong experience if they have large instability.

I understand the code branching argument for having separate flags, this is something that is very painful in existing tools like Babel and Webpack, how plugins and options interact. But I think here it's probably worthwhile. If we use that as the main argument then we're likely to reject any small convenience options in the future.

**Reply by ematipico** [reply]
It seems that the main concerns keeps being about "the users" while the focus should shift towards *us maintainers*. 

So, my main proposal of a single flag was inspired by Rust, which seems to be working quite well and they can do so by leveraging the language itself. 

Now, I see the point of many flags VS one flag, because "people don't know what they are getting into" or "what kind of experimental features I am actually opting-in" arguments are really valid and reasonable, and they can't be compared to Rust because in that case, people need to actually import possible unstable functions, so users see what the are using.

This point is not valid for us because we use a CLI/LSP to opt-in into these experimental features. There's a point here. 

Although, we know that **objectively** maintaining more stuff (in this case options) would take more time to the team (imagine handling the option `--experimental-new-way-of-format` when called with the command `rome install`, we need to handle errors, diagnostics, etc.) as opposed to one option, and the team should not bother spending too much time on maintaining code that **will eventually be deleted anyways**.

## Counter proposal

- maintain one single option for opt-in the experimental features
- create a new section in our website where the team will list all the experimental features that we are currently developing
- disclaim the `--experimental`* option as **not production fit** and only for development purpose, to help the team to fix/polish WIP features
- remove these experimental features from said section once they are shipped as ready

**Reply by yassere** [reply]
> my main proposal of a single flag was inspired by Rust

I don't understand what you mean by this. Rust has a structured process for adding new features behind individual [feature gates](https://rustc-dev-guide.rust-lang.org/implementing_new_features.html#stability-in-code) and then another process for [stabilizing features](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html).

Using the `nightly` channel doesn't automatically enable all of those unstable features, but it allows you to opt-in to them individually.

**Reply by ematipico** [reply]
> > my main proposal of a single flag was inspired by Rust
> 
> I don't understand what you mean by this. Rust has a structured process for adding new features behind individual [feature gates](https://rustc-dev-guide.rust-lang.org/implementing_new_features.html#stability-in-code) and then another process for [stabilizing features](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html).
> 
> Using the `nightly` channel doesn't automatically enable all of those unstable features, but it allows you to opt-in to them individually.

I had a chat with Leo about too, and you can opt-in unstable features but just using an environment variable.

Essentially, it all boils down to a single feature flag. Either way, let's bring the focus on the main proposal, wether I'm wrong or not, it should not affect our decision.

**Reply by leops** [reply]
I think there's a bit of a confusion here, setting the `RUSTC_BOOTSTRAP` environment variable does not actually enable experimental features, it only allows them to be enabled on stable builds (they can always be enabled on nightly builds). Each feature still has to be turned on individually whether it's a compiler-level feature (by setting a command line flag `-Z macro-backtrace`) or a language-level feature (by using a crate-level attribute like `#![feature(generic_associated_types)]`)

My point in bringing up this environment variable was more about the stability commitment that goes into these experimental features: most users should never use `RUSTC_BOOTSTRAP`, but the fact that this code ships even in stable builds means they must be designed so they don't impact the normal working of the compiler is they are not enabled

**Reply by MichaReiser** [reply]
This is an excellent point @leops and it highlights to me that we may want to try different solutions. 

That's why it could make sense to clarify what we are trying to solve by introducing experimental features:

1. Being able to ship stable Rome releases where unfinished features are disabled without the need for feature branches
2. Allow users to use unfinished features 

In my view, we're mainly solving for 1 which is why I like what @leops suggests. 

Allowing users to use experimental (or even unfinished) features raises a lot of questions as we've seen and it would be good if we can defer them until later. 

Edit: 

To clarify my terminology 

* unfinished feature: A feature that's still in development and isn't ready to be used in any context. For example, the JSX formatting. 
* experimental feature: An implemented feature that is ready to be used but it's undecided yet if it's something Rome should support in the long term or some details on the feature are unclear/undecided

**Reply by ematipico** [reply]
>1.  Being able to ship stable Rome releases where unfinished features are disabled without the need for feature branches

Exactly. That's what I have been trying to suggest and focus (maybe I haven't done a good job). If users won't be able to use the experiments, even better. 

I want to suggest some peace of mind to us when creating new stuff. Take this PR for example https://github.com/rome/tools/pull/2402, I would like to merge it without affecting the normal behaviour of the formatter. Once the sorting is finished (it will take more than one PR), we remove whatever flag we have and we will have the sorting import ready.

This would allow us to make smaller PRs, better review process, gradual testing, etc. Imagine creating a new feature in one single PR... sometimes that doesn't work.

**Reply by yassere** [reply]
One of my main concerns is that if multiple new features all make widespread changes behind `experimental` checks, it can become difficult to disentangle those changes when it’s time to un-gate and stabilize a feature. If we want to have a single flag, then we need to have strict guidelines for how and where you can introduce an `experimental` check.

And if we don’t want to end up with developers adding ad-hoc individual feature flags, we’ll need to be diligent about restricting the introduction of new config settings.

> What about if features have individual flags but Rome only supports switching them all ON or OFF?

This proposal from Micha would solve my concern about disentangling features. And it would be easier to convert into a complete feature flag system later on. Although I'd still prefer that features be individually opt-in from the start. One of the major advantages of feature flags is that adding a new gated feature should never break anyone's existing usage. We lose that advantage if people are using an `all_features` flag.

> How do we ensure that all experimental features all play nicely together?

We could certainly require that our tests pass with all features enabled, but I don't think we need to worry about every unique combination of unstable features. When a feature is stabilized, we'll ensure it plays nicely with every other stable feature.

> (imagine handling the option --experimental-new-way-of-format when called with the command rome install, we need to handle errors, diagnostics, etc.) as opposed to one option

I'm confused about this, and maybe this is the main source of our disconnect. I was imagining that individual commands wouldn't need to do any special handling of feature flags that they don't specifically care about. The "central crate" you mentioned would populate a non-exhaustive `unstable_features` and that config would be available throughout most of the codebase.

So in your example, the `rome install` command wouldn't need to know that `new-way-of-format` even exists as long as the formatter API doesn't change, but if it calls out internally to the formatter, then the formatter itself can check for that feature flag.

But there's certainly a maintenance burden involved with allowing feature flags to permeate our codebase.

Is the proposal instead that `experimental` code paths would be confined to the periphery and primarily in the command-handling itself?  So it would mostly be that each command has a single "experimental" version of itself?

**Reply by ematipico** [reply]
@yassere 

>This proposal from Micha would solve my concern about disentangling features. And it would be easier to convert into a complete feature flag system later on. Although I'd still prefer that features be individually opt-in from the start. One of the major advantages of feature flags is that adding a new gated feature should never break anyone's existing usage. We lose that advantage if people are using an all_features flag.

Could expand why, you, **as developer**, would like have multiple flags? Users won't need to use these features and they don't have too. If possible, we will strip out any logic around enabling them in production releases.

My main focus, as developer, is to understand your needs so I can make a proposal for us as a team.

**Reply by MichaReiser** [reply]
Having multiple flags for different features is helpful once you stabilize a feature. It's very hard to find all branching points if the conditions don't use a unique identifier. Mainly, it allows use to grep for all the relevant use cases to easily remove it, remove the gating, etc. 

> Although I'd still prefer that features be individually opt-in from the start.

 I would like that too but I think we should try to spend as little time as possible because the only reason we want it right now is to unblock releases. I rather want to avoid turning this into its own multi-week project. 
 
I'm a bit worried that we'll start arguing forth and back. My main concern is that we should focus on a minimal solution that allows us shipping the next releases without enabling unfinished features. I don't need a way for users to opt-in but if they can, then I don't mind much. I'm happy with whatever solution delivers that.

**Reply by yassere** [reply]
> Could expand why, you, as developer, would like have multiple flags?

My primary concern is the same as Micha's about the ease of removing conditional code paths when we stabilize a feature. I never want anyone to look at a piece of "experimental" code and wonder if it's part of feature "foo" or feature "bar". Having unique names for feature gates feels much more robust than relying on comments.

> Take this PR for example https://github.com/rome/tools/pull/2402, I would like to merge it without affecting the normal behaviour of the formatter.

Now that I've looked at this draft PR, I think I'm starting to understand the source of my confusion.
This PR already contains feature toggles:

```rs
/// Data structure that holds information of possible experimental features inside the formatter
///  
/// If an unstable feature becomes stable, it should be removed from there and the new logic should
/// become the norm
#[derive(Debug, Default, Clone, Copy)]
pub struct FormatUnstableFeatures {
    /// A feature to sort imports in a consistent way
    pub unstable_sort_imports: bool,
}
```

If you want to replace the `unstable_sort_imports` check in that PR with a non-specific `experimental` check, I think that would be confusing and hard to maintain.

But if not, then we _would_ have individual feature flags (whether or not they're  independently toggle-able via CLI). And if we don't want a single unified approach for feature flags, then it sounds like we'd end up with each crate implementing its own version. Which is fine for now.

@ematipico Is this closer to your actual proposal or am I still misunderstanding:

There's a single `experimental` flag that's _only_ checked from a command handler (or LSP) and each command (or LSP handler) decides what "experimental" means for that command.

So when the `format` command receives an `experimental` flag, it knows to pass `unstable_sort_imports` when formatting. But the `rome_js_formatter` itself would never have access to a flag called `experimental`.

**Reply by MichaReiser** [reply]
I'm not sure if it's possible to completely isolate changes. Let's say we change how range formatting works and must changes in the lsp and formatter crate. Ideally, both crates test for the same feature flag.

A very simple implementation could be a shared crate that allows testing if a specific feature (feature enum?).

This crate must be initialized during startup or all features are disabled

**Reply by yassere** [reply]
I created draft PR #2487 as a simple example of feature flags to help with this discussion. I'll just close it if we decide against that sort of approach or if someone wants to work on an alternate implementation.

**Reply by ematipico** [reply]
I would have preferred a counter proposal instead of jumping straight to the code.

Anyway, from your draft PR, could you propose how and if us developers should document the features that are in WIP?

Should user able to opt-in these flags with the release version of the binary?

**Reply by yassere** [reply]
> I would have preferred a counter proposal instead of jumping straight to the code.

The code is the counter proposal. It seemed easier than going back and forth with comments, and we can always just close the draft PR if we don’t like it.

> could you propose how and if us developers should document the features that are in WIP?

I modified the macro to allow for adding a doc comment that shows up whenever you hover over any usage of the flag:
```rs
declare_feature_flags!(
    /// A new way of linebreaking
    new_linebreaking,
    /// A new way of spacing
    new_spacing
);
```
There's more we can borrow from Rust's [flags](https://github.com/rust-lang/rust/blob/d8e59edbfa47ff38e23e6dedab6bedd3b41895e0/compiler/rustc_feature/src/active.rs#L127) like documenting the issue associated with the feature.
In our case, I think that feature flags should be temporary and deleted as soon as possible rather than documenting them long-term.

> Should user able to opt-in these flags with the release version of the binary?

If we're primarily intending these feature flags as a way to merge unfinished code without disrupting releases, then we should discourage their usage by consumers (and maybe print appropriate warnings). But it can still be useful to have them available in the release binary for benchmarking and testing.

**Reply by ematipico** [reply]
Sorry @yassere, I should have expanded. With documenting I meant what I said in my counter proposal. Having a MD document where to track the unfinished features. Should we? Are comments enough? Or Do you think the `rome_feature` will be enough on the long run?


Overall, I like the proposal!

**Reply by yassere** [reply]
Going back to the distinction that Micha was making:
> 1. Being able to ship stable Rome releases where unfinished features are disabled without the need for feature branches
> 2. Allow users to use unfinished features

Since we're currently only solving case 1, I think that comments are sufficient documentation.
I agree that case 2 should be deferred and we'd likely want better documentation for those flags. If we decide to pursue this proposal, we'll want to revise my PR to make it clear that those are "unfinished" feature flags. If we later add "experimental" feature flags for public usage, they should be clearly differentiated.

> Overall, I like the proposal!

Thanks! I think it's mostly just what @MichaReiser was suggesting above.
I haven't put much thought yet into the details of the implementation itself.

**Reply by ematipico** [reply]
The PR was merged, which means that the team as arrived at a consensus. Thank you team!
